client/src/details/SFSClientImpl.h

```diff
@@ -97,9 +98,9 @@ class SFSClientImpl : public SFSClientInterface
     void SetCustomBaseUrl(std::string customBaseUrl);
 
     /**
-     * @return The URL for the SFS service based on the parameters passed to the constructor
+     * @return A SFSUrlBuilder object that can be used to build URLs for the SFS service
      */
-    std::string GetBaseUrl() const;
+    std::unique_ptr<SFSUrlBuilder> MakeUrlBuilder() const;

```

Comment:
SFSUrlBuilder is non-copyable but you should be able to return-by-value here due to return value optimization (RVO). Did you try it?

Reply:
The issue was in SFSUrlBuilder::CreateFromAccountId() and SFSUrlBuilder::CreateFromCustomUrl().
I couldn't return by value there because NRVO was not allowed since SFSUrlBuilder is non-copyable

```
SFSUrlBuilder CreateFromAccountId(...)
{
    SFSUrlBuilder builder(std::move(instanceId), std::move(nameSpace), handler);
    builder.SetScheme(Scheme::Https);
    builder.SetHost(accountId + "." + std::string(c_apiDomain));
    return builder;
}
```

But it seems that creating a private constructor and switching that to below works

```
SFSUrlBuilder CreateFromAccountId(...)
{
    return SFSUrlBuilder(accountId, std::move(instanceId), std::move(nameSpace), handler);
}
```

I had to create two factories because the arguments are the same types and the constructors wouldn't be distinguishable. Another way to handle this is to create a fake type only to differentiate the constructors.

Reply:
Create a wrapper struct to facilitate this